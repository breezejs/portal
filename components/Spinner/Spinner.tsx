import {darken} from 'polished';
import React, {Fragment, memo} from 'react';
import styled from 'styled-components';
import {GlobalStylesTheme} from '../../core/GlobalStyles';

const SpinnerElement = styled.div`
  @keyframes spinner {
    to {
      transform: rotate(360deg);
    }
  }

  height: 100px;
  left: 50%;
  margin: -50px 0 0 -50px;
  position: fixed;
  top: 50%;
  width: 100px;
  z-index: ${GlobalStylesTheme.zIndexSpinner};

  ::before {
    content: "Loading...";
    height: 100px;
    left: 50%;
    margin: -50px 0 0 -50px;
    position: absolute;
    top: 50%;
    width: 100px;
  }

  :not(:required) {
    ::before {
      animation: spinner .6s linear infinite;
      border: 5px solid ${darken(0.1, GlobalStylesTheme.bodyBg)};
      border-radius: 50%;
      border-top-color: ${GlobalStylesTheme.success};
      content: "";
    }
  }
`;

interface ISpinnerProps {
  show: boolean;
}

function Spinner ({show}: ISpinnerProps) {
  function renderSpinner () {
    return (
      <SpinnerElement />
    );
  }

  return (
    <Fragment>{show && renderSpinner()}</Fragment>
  );
}

export default memo(Spinner);
