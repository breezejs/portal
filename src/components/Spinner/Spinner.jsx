import {darken} from 'polished';
import PropTypes from 'prop-types';
import React, {memo} from 'react';
import styled from 'styled-components';
import {GlobalStylesTheme} from '../GlobalStyles';

const SpinnerElement = styled.div`
  @keyframes spinner {
    to {
      transform: rotate(360deg);
    }
    
  }

  height: 100px;
  left: 50%;
  margin: -50px 0 0 -50px;
  position: fixed;
  top: 50%;
  width: 100px;
  z-index: ${GlobalStylesTheme.zIndexSpinner};

  ::before {
    content: "Loading...";
    height: 100px;
    left: 50%;
    margin: -50px 0 0 -50px;
    position: absolute;
    top: 50%;
    width: 100px;
  }

  :not(:required) {
    ::before {
      animation: spinner .6s linear infinite;
      border: 5px solid ${darken(0.1, GlobalStylesTheme.bodyBg)};
      border-radius: 50%;
      border-top-color: ${({theme}) => GlobalStylesTheme[theme]};
      content: "";
    }
  }
`;

function Spinner ({theme}) {
  return (
    <SpinnerElement theme={theme} />
  );
}

Spinner.propTypes = {
  theme: PropTypes.oneOf(['danger', 'info', 'primary', 'secondary', 'success', 'warning'])
};

export default memo(Spinner);
