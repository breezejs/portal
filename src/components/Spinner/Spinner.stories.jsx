import {optionsKnob, withKnobs} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import React from 'react';
import GlobalStyles from '../GlobalStyles';
import Spinner from './Spinner';

const stories = storiesOf('Spinner', module);

const options = {
  danger: 'danger',
  info: 'info',
  primary: 'primary',
  secondary: 'secondary',
  success: 'success',
  warning: 'warning'
};
const optionsObj = {
  display: 'select'
};

stories
  .addDecorator(withKnobs)
  .addDecorator((storyFn) => (
    <div>
      <GlobalStyles />
      {storyFn()}
    </div>
  ))
  .add('with multiple themes', () => (
    <Spinner theme={optionsKnob('Theme', options, 'info', optionsObj)} />
  ));
