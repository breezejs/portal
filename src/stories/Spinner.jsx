import {storiesOf} from '@storybook/react';
import React from 'react';
import GlobalStyles from '../components/GlobalStyles';
import Spinner from '../components/Spinner';

const stories = storiesOf('Spinner', module);

stories
  .addDecorator((storyFn) => (
    <div>
      <GlobalStyles />
      {storyFn()}
    </div>
  ))
  .add('with primary theme', () => (
    <Spinner theme="primary" />
  ))
  .add('with secondary theme', () => (
    <Spinner theme="secondary" />
  ))
  .add('with info theme', () => (
    <Spinner theme="info" />
  ))
  .add('with success theme', () => (
    <Spinner theme="success" />
  ))
  .add('with danger theme', () => (
    <Spinner theme="danger" />
  ))
  .add('with warning theme', () => (
    <Spinner theme="warning" />
  ));
