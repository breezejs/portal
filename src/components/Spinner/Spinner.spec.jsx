import {shallow} from 'enzyme';
import React from 'react';
import Spinner from './Spinner';

describe('Spinner', () => {
  describe('with no theme set', () => {
    it('should render and default to info', () => {
      const result = shallow(
        <Spinner />
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('with primary theme', () => {
    it('should render', () => {
      const result = shallow(
        <Spinner theme='primary' />
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('with secondary theme', () => {
    it('should render', () => {
      const result = shallow(
        <Spinner theme='secondary' />
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('with success theme', () => {
    it('should render', () => {
      const result = shallow(
        <Spinner theme='success' />
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('with danger theme', () => {
    it('should render', () => {
      const result = shallow(
        <Spinner theme='danger' />
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('with warn theme', () => {
    it('should render', () => {
      const result = shallow(
        <Spinner theme='warn' />
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('with info theme', () => {
    it('should render', () => {
      const result = shallow(
        <Spinner theme='info' />
      );

      expect(result).toMatchSnapshot();
    });
  });
});
