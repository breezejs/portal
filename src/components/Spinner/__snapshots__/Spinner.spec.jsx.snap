// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner with danger theme should render 1`] = `
<styled.div
  theme="danger"
/>
`;

exports[`Spinner with info theme should render 1`] = `
<styled.div
  theme="info"
/>
`;

exports[`Spinner with no theme set should render and default to info 1`] = `
<styled.div
  theme="info"
/>
`;

exports[`Spinner with primary theme should render 1`] = `
<styled.div
  theme="primary"
/>
`;

exports[`Spinner with secondary theme should render 1`] = `
<styled.div
  theme="secondary"
/>
`;

exports[`Spinner with success theme should render 1`] = `
<styled.div
  theme="success"
/>
`;

exports[`Spinner with warn theme should render 1`] = `
<styled.div
  theme="warn"
/>
`;
